File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaEnumDataType.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.sdk.core.ValueRank;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.UaStructuredType;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.structured.EnumValueType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManager;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaDataTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.types.common.UaStructureUtilities;
import org.springframework.lang.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class UaEnumDataType extends UaDataType {
    public UaEnumDataType(
            String dataTypeId,
            LocalizedText displayName,
            List<String> enumStrings,
            NodeManager nodeManager)
    {
        super(new NodeId(nodeManager.nsIndex(), dataTypeId),
                dataTypeId,
                displayName,
                false);

        setParentType(UaDataTypes.Enumeration);

        String enumStringsId = dataTypeId;
        enumStringsId += "-EnumStrings";

        UaVariable enumStringsVariable = setEnumStrings(
                new NodeId(nodeManager.nsIndex(), enumStringsId),
                enumStrings);

        nodeManager.addNode(this);
        nodeManager.addNode(enumStringsVariable);
    }

    public UaEnumDataType(
            String dataTypeId,
            LocalizedText displayName,
            EnumValueType[] enumValues,
            NodeManager nodeManager)
    {
        super(new NodeId(nodeManager.nsIndex(), dataTypeId),
                dataTypeId,
                displayName,
                false);

        setParentType(UaDataTypes.Enumeration);

        String enumValuesId = dataTypeId;
        enumValuesId += "-EnumValues";

        UaVariable enumValuesVariable = setEnumValues(
                new NodeId(nodeManager.nsIndex(), enumValuesId),
                Arrays.asList(enumValues));

        nodeManager.addNode(this);
        nodeManager.addNode(enumValuesVariable);
    }

    public @Nullable Integer parse(Object value)
    {
        return null;
    }

    private UaVariable setEnumStrings(
            NodeId nodeId,
            List<String> enumStrings)
    {
        UaVariable variable = new UaVariable(
                nodeId,
                "EnumStrings",
                new LocalizedText("EnumStrings"),
                NodeIds.LocalizedText,
                ValueRank.OneDimension.getValue(),
                AccessLevel.CurrentRead.getValue(),
                UaVariableTypes.PropertyType);

        LocalizedText[] enumString_LText = new LocalizedText[enumStrings.size()];
        for (int i = 0; i<enumStrings.size(); ++i)
        {
            enumString_LText[i] = new LocalizedText(enumStrings.get(i));
        }

        variable.setValue(new Variant(enumString_LText));
        addMemberNode(variable);
        return variable;
    }

    private UaVariable setEnumValues(
            NodeId nodeId,
            List<EnumValueType> enumValues)
    {
        UaVariable variable = new UaVariable(
                nodeId,
                "EnumValues",
                new LocalizedText("EnumValues"),
                NodeIds.EnumValueType,
                ValueRank.OneDimension.getValue(),
                AccessLevel.CurrentRead.getValue(),
                UaVariableTypes.PropertyType);

        List<UaStructuredType> structs = new ArrayList<>(enumValues);
        variable.setValue(UaStructureUtilities.toVariant(structs));

        addMemberNode(variable);
        return variable;
    }
}

File: opcua-webservice-node/source/src/server/addressspace/nodes/UaEnumDataType.ts
import {
        UaAccessLevel,
        UaExtensionObject,
        UaEnumValueType,
        UaLocalizedText,
        UaNodeId,
        UaValueRank,
        UaVariant,
} from "opcua-webapi-ts";
import { NodeManager } from "../nodemanager";
import { UaDataTypes, UaVariableTypes } from "./builtin";
import { UaDataType } from "./UaDataType";
import { UaVariable } from "./UaVariable";

export class UaEnumDataType extends UaDataType {
        private static readonly EnumStringsBrowseName = "EnumStrings";
        private static readonly EnumValuesBrowseName = "EnumValues";

        constructor(
                dataTypeId: string,
                displayName: UaLocalizedText,
                enumStrings: Array<string>,
                nodeManager: NodeManager
        );
        constructor(
                dataTypeId: string,
                displayName: UaLocalizedText,
                enumValues: Array<UaEnumValueType>,
                nodeManager: NodeManager
        );
        constructor(
                dataTypeId: string,
                displayName: UaLocalizedText,
                enumMembers: Array<string> | Array<UaEnumValueType>,
                nodeManager: NodeManager
        ) {
                super(
                        new UaNodeId(dataTypeId, nodeManager.nsIndex()),
                        dataTypeId,
                        displayName,
                        false,
                );

                this.setParentType(UaDataTypes.Enumeration);

                let enumVariable: UaVariable;
                if (this.isEnumValueTypeArray(enumMembers)) {
                        enumVariable = this.setEnumValues(
                                new UaNodeId(`${dataTypeId}-${UaEnumDataType.EnumValuesBrowseName}`, nodeManager.nsIndex()),
                                enumMembers as Array<UaEnumValueType>,
                        );
                } else {
                        enumVariable = this.setEnumStrings(
                                new UaNodeId(`${dataTypeId}-${UaEnumDataType.EnumStringsBrowseName}`, nodeManager.nsIndex()),
                                enumMembers as Array<string>,
                        );
                }

                nodeManager.addNode(this);
                nodeManager.addNode(enumVariable);
        }

        parse(value: unknown): number | null {
                return null;
        }

        private isEnumValueTypeArray(values: Array<string> | Array<UaEnumValueType>): boolean {
                return values.length > 0 && typeof values[0] !== "string";
        }

        private setEnumStrings(nodeId: UaNodeId, enumStrings: Array<string>): UaVariable {
                const variable = this.createPropertyVariable(
                        nodeId,
                        UaEnumDataType.EnumStringsBrowseName,
                        UaDataTypes.LocalizedText.nodeId,
                );

                const localizedEnumStrings: Array<UaLocalizedText> = [];
                for (const enumString of enumStrings) {
                        localizedEnumStrings.push(new UaLocalizedText(enumString));
                }

                variable.value = UaVariant.localizedTexts(localizedEnumStrings);
                this.addMemberNode(variable);

                return variable;
        }

        private setEnumValues(nodeId: UaNodeId, enumValues: Array<UaEnumValueType>): UaVariable {
                const variable = this.createPropertyVariable(
                        nodeId,
                        UaEnumDataType.EnumValuesBrowseName,
                        UaDataTypes.EnumValueType.nodeId,
                );

                const extensionObjects: Array<UaExtensionObject> = [];
                for (const enumValue of enumValues) {
                        extensionObjects.push(enumValue.toExtensionObject());
                }

                variable.value = UaVariant.extensionObjects(extensionObjects);
                this.addMemberNode(variable);

                return variable;
        }

        private createPropertyVariable(nodeId: UaNodeId, browseName: string, dataType: UaNodeId): UaVariable {
                return new UaVariable(
                        nodeId,
                        browseName,
                        new UaLocalizedText(browseName),
                        dataType,
                        UaValueRank.OneDimension,
                        UaAccessLevel.CurrentRead,
                        UaVariableTypes.PropertyType,
                );
        }
}